test/Exchange.spec.ts

```diff
@@ -0,0 +1,40 @@
+import { expect } from "chai";
+import hre, { deployments, waffle } from "hardhat";
+import "@nomiclabs/hardhat-ethers";
+
+describe("Exchange", async () => {
+  const [user1] = waffle.provider.getWallets();
+
+  const setupTests = deployments.createFixture(async ({ deployments }) => {
+    await deployments.fixture();
+    const Pool = await hre.ethers.getContractFactory("TestUniswapPool");
+    const pool = await Pool.deploy(
+      ["8094772033607", "8094774061327"],
+      [
+        "198044337547992734631384961681122638580576",
+        "198044337547992734631700968341745576827435",
+      ]
+    );
+    const Factory = await hre.ethers.getContractFactory("TestUniswapFactory");
+    const factory = await Factory.deploy(pool.address);
+    const Token = await hre.ethers.getContractFactory("TestToken");
+    const token0 = await Token.deploy("Token0", "Token0", 6);
+    const token1 = await Token.deploy("Token1", "Token1", 18);
+    const Exchange = await hre.ethers.getContractFactory("Exchange");
+    const exchange = await Exchange.deploy();
+    await exchange.initialize(user1.address);
+    await exchange.setUp(factory.address, 3000, token0.address);
+    return {
+      exchange,
+      token0,
+      token1,
+    };
+  });
+
+  describe("exchangeRateOf()", () => {
+    it("should return usd rate", async () => {
+      const { exchange, token1 } = await setupTests();
+      expect((await exchange.exchangeRateOf(token1.address)).value).to.be.gt(0);
```

Comment:
could we have some more precise tests? we have a ton of very sophisticated math lib contracts that are using assembly, it would be nice to see some actual rates being derived from various liquidity positions. let's just use some nice round numbers to test the different rates.

Reply:
I've updated the tests.